src/nvim/screen.c

```diff
@@ -2073,6 +2073,212 @@ fill_foldcolumn (
+      shl_flag = true;
+    } else
+      shl = &cur->hl;
+    if (if_col == (long)shl->startcol)
+      attr = shl->attr;
+    if (shl != &search_hl && cur != NULL)
+      cur = cur->next;
+  }
+  return attr;
+}
+
+static int get_search_attr_cur(win_T* wp) {
+  int attr = search_hl.attr_cur;
+  bool shl_flag = false;
+  matchitem_T *cur = wp->w_match_head;
+  match_T     *shl;
+  while (cur != NULL || shl_flag == false) {
+    if (shl_flag == false && ((cur != NULL && cur->priority > SEARCH_HL_PRIORITY) || cur == NULL)) {
+      shl = &search_hl;
+      shl_flag = true;
+    } else
+      shl = &cur->hl;
+    if (shl->attr_cur != 0)
+      attr = shl->attr_cur;
+    if (shl != &search_hl && cur != NULL)
+      cur = cur->next;
+  }
+  return attr;
+}
+
+static bool init_syntax(int lnum, win_T* wp) {
+  int save_did_emsg;
+  if (syntax_present(wp) && !wp->w_s->b_syn_error) {
+    save_did_emsg = did_emsg;
+    did_emsg = FALSE;
+    syntax_start(wp, lnum); // Prepare for syntax highlighting in this line.
+    if (did_emsg) {         // On error, stop syntax highlighting.
+      wp->w_s->b_syn_error = TRUE;
+    } else {
+      did_emsg = save_did_emsg;
+      return true;
+    }
+  }
+  return false;
+}
+
+struct VisualPos {
+  int from; int to;
+  bool in_visual;
+  pos_T pos;
+};
+
+static struct VisualPos init_visual(int lnum, win_T* wp) {
+  struct VisualPos res = {
+    .from = -10, .to = MAXCOL,
+    .in_visual = false,
+    .pos = { 0 }
+  };
+
+  pos_T *top, *bot;
+  if (ltoreq(curwin->w_cursor, VIsual)) {
+    top = &curwin->w_cursor;
+    bot = &VIsual;
+  } else {                          /* Visual is before curwin->w_cursor */
+    top = &VIsual;
+    bot = &curwin->w_cursor;
+  }
+
+  if (lnum >= top->lnum && lnum <= bot->lnum)
+    res.in_visual = true;
+
+  if (VIsual_mode == Ctrl_V) {
+    if (res.in_visual) { /* block mode */
+      res.from = wp->w_old_cursor_fcol;
+      res.to = wp->w_old_cursor_lcol;
+    }
+  } else { /* non-block mode */
+    if (lnum > top->lnum && lnum <= bot->lnum) {
+      res.from = 0;
+    } else if (lnum == top->lnum) {
+      if (VIsual_mode == 'V') { /* linewise */
+        res.from = 0;
+      } else {
+        getvvcol(wp, top, (colnr_T *)&res.from, NULL, NULL);
+        if (gchar_pos(top) == NUL)
+          res.to = res.from + 1;
+      }
+    }
+    if (VIsual_mode != 'V' && lnum == bot->lnum) {
+      if (*p_sel == 'e' && bot->col == 0 && bot->coladd == 0 )
+        { res.from = -10; res.to = MAXCOL; }
+      else if (bot->col == MAXCOL)
+        { res.to = MAXCOL; }
+      else {
+
+        res.pos = *bot;
+        if (*p_sel == 'e') {
+          getvvcol(wp, &(res.pos), (colnr_T *)&res.to, NULL, NULL);
+        } else {
+          getvvcol(wp, &(res.pos), NULL, NULL, (colnr_T *)&res.to);
+          ++res.to;
+        }
+      }
+    }
+  }
+
+  return res;
+}
+
+#define MB_IS_WIDE() (has_mbyte && (*mb_char2cells)(mb_c) > 1)
+
+static void write_space (unsigned off) {
```

Comment:
is this really so common it deserves a separate function? if it is, why not make it `write_ascii(off, char)` ?